pkg/eventfilter/subscriptionsapi/all_filter.go

```diff
@@ -18,34 +18,88 @@ package subscriptionsapi
 
 import (
 	"context"
+	"sync"
 
 	cloudevents "github.com/cloudevents/sdk-go/v2"
+	"go.uber.org/atomic"
 	"go.uber.org/zap"
 	"knative.dev/pkg/logging"
 
 	"knative.dev/eventing/pkg/eventfilter"
 )
 
-type allFilter []eventfilter.Filter
+type filterCount struct {
+	filter eventfilter.Filter
+	count  atomic.Uint64
+}
+
+type allFilter struct {
+	filters []filterCount
+	rwMutex sync.RWMutex
+	c       chan int
+	d       chan bool
+}
 
 // NewAllFilter returns an event filter which passes if all the contained filters pass
 func NewAllFilter(filters ...eventfilter.Filter) eventfilter.Filter {
-	return append(allFilter{}, filters...)
+	filterCounts := make([]filterCount, len(filters))
+	for i, filter := range filters {
+		filterCounts[i] = filterCount{
+			count:  *atomic.NewUint64(uint64(0)),
+			filter: filter,
+		}
+	}
+	f := &allFilter{
+		filters: filterCounts,
+		c:       make(chan int),
+		d:       make(chan bool),
+	}
+	go f.optimizeLoop()
+	return f
 }
 
-func (filter allFilter) Filter(ctx context.Context, event cloudevents.Event) eventfilter.FilterResult {
+func (filter *allFilter) Filter(ctx context.Context, event cloudevents.Event) eventfilter.FilterResult {
 	res := eventfilter.NoFilter
 	logging.FromContext(ctx).Debugw("Performing an ALL match ", zap.Any("filters", filter), zap.Any("event", event))
-	for _, f := range filter {
-		res = res.And(f.Filter(ctx, event))
+	filter.rwMutex.RLock()
+	defer filter.rwMutex.RUnlock()
+	for i, f := range filter.filters {
+		res = res.And(f.filter.Filter(ctx, event))
 		// Short circuit to optimize it
 		if res == eventfilter.FailFilter {
+			filter.c <- i
```

Comment:
I'm worried about this could block if the background gorouting is not listening at that point, I think we need buffering and avoid blocking when the buffer is full